src/beast/beast/crypto/impl/sha512_context.h

```diff
@@ -0,0 +1,276 @@
+                        w[j - 16];
+        for (j = 0; j < 8; j++)
+            wv[j] = ctx.h[j];
+        for (j = 0; j < 80; j++) {
+            t1 = wv[7] + BEAST_SHA512_F2(wv[4]) +
+                BEAST_SHA2_CH(wv[4], wv[5], wv[6]) +
+                    K[j] + w[j];
+            t2 = BEAST_SHA512_F1(wv[0]) +
+                BEAST_SHA2_MAJ(wv[0],
+                    wv[1], wv[2]);
+            wv[7] = wv[6];
+            wv[6] = wv[5];
+            wv[5] = wv[4];
+            wv[4] = wv[3] + t1;
+            wv[3] = wv[2];
+            wv[2] = wv[1];
+            wv[1] = wv[0];
+            wv[0] = t1 + t2;
+        }
+        for (j = 0; j < 8; j++)
+            ctx.h[j] += wv[j];
+    }
+}
+
+template <class = void>
+void init (sha512_context& ctx) noexcept
+{
+    ctx.h[0] = 0x6a09e667f3bcc908ULL;
+    ctx.h[1] = 0xbb67ae8584caa73bULL;
+    ctx.h[2] = 0x3c6ef372fe94f82bULL;
+    ctx.h[3] = 0xa54ff53a5f1d36f1ULL;
+    ctx.h[4] = 0x510e527fade682d1ULL;
+    ctx.h[5] = 0x9b05688c2b3e6c1fULL;
+    ctx.h[6] = 0x1f83d9abfb41bd6bULL; 
+    ctx.h[7] = 0x5be0cd19137e2179ULL;
+    ctx.len = 0;
+    ctx.tot_len = 0;
+}
+
+template <class = void>
+void update (sha512_context& ctx,
+    void const* message, std::size_t size) noexcept
+{
+    auto const pm = reinterpret_cast<
+        unsigned char const*>(message);
+    unsigned int block_nb;
+    unsigned int new_len, rem_len, tmp_len;
+    const unsigned char *shifted_message;
+    tmp_len = sha512_context::block_size - ctx.len;
+    rem_len = size < tmp_len ? size : tmp_len;
+    std::memcpy(&ctx.block[ctx.len], pm, rem_len);
+    if (ctx.len + size < sha512_context::block_size) {
+        ctx.len += size;
+        return;
+    }
+    new_len = size - rem_len;
+    block_nb = new_len / sha512_context::block_size;
+    shifted_message = pm + rem_len;
+    sha512_transform(ctx, ctx.block, 1);
+    sha512_transform(ctx, shifted_message, block_nb);
+    rem_len = new_len % sha512_context::block_size;
+    std::memcpy(ctx.block, &shifted_message[
+        block_nb << 7], rem_len);
+    ctx.len = rem_len;
+    ctx.tot_len += (block_nb + 1) << 7;
+}
+
+template <class = void>
+void finish (sha512_context& ctx,
+    void* digest) noexcept
+{
+    auto const pd = reinterpret_cast<
+        unsigned char*>(digest);
+    unsigned int block_nb;
+    unsigned int pm_len;
+    unsigned int len_b;
+    int i;
+    block_nb = 1 + ((sha512_context::block_size - 17) <
+        (ctx.len % sha512_context::block_size));
+    len_b = (ctx.tot_len + ctx.len) << 3;
+    pm_len = block_nb << 7;
+    memset(ctx.block + ctx.len, 0, pm_len - ctx.len);
+    ctx.block[ctx.len] = 0x80;
+    BEAST_SHA2_UNPACK32(len_b, ctx.block + pm_len - 4);
+    sha512_transform(ctx, ctx.block, block_nb);
+    for (i = 0 ; i < 8; i++)
+        BEAST_SHA2_UNPACK64(ctx.h[i], &pd[i << 3]);
+}
+
+template <class = void>
+void secure_erase (sha512_context& ctx)
+{
+    std::memset(ctx.block, 0, sizeof(ctx.block));
```

Comment:
David answers this question:
http://stackoverflow.com/questions/15538366/can-memset-function-call-be-removed-by-compiler